### PR TITLE
chore: reduce refactor

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

File: pkg/reducer/examples/counter/main.go [hunk 1]
```diff
@@ -20,5 +20,7 @@ func reduceCounter(_ context.Context, keys []string, reduceCh <-chan reducer.Dat
 }
 
 func main() {
-	reducer.NewServer(reducer.ReducerFunc(reduceCounter)).Start(context.Background())
+	reducer.NewServer(reducer.CreateReducerFunc(func() reducer.Reducer {
```
Comment (vigith) [inline pkg/reducer/examples/counter/main.go:23]:
```suggestion
	reducer.NewServer(reducer.CreateReducerFn(func() reducer.Reducer {
```

File: pkg/reducer/interface.go [hunk 2]
```diff
@@ -23,11 +23,25 @@ type IntervalWindow interface {
...
 	Reduce(ctx context.Context, keys []string, reduceCh <-chan Datum, md Metadata) Messages
 }
 
+// CreateReducerFunc is a utility type used to convert a create function to a CreateReducer type.
+type CreateReducerFunc func() Reducer
+
+// Create implements the function of CreateReducer.
+func (rf CreateReducerFunc) Create() Reducer {
+	return rf()
+}
```
Comment (vigith) [inline pkg/reducer/interface.go:43]:
```suggestion
// CreateReducerFn is a utility type used to convert a create function to a CreateReducer type.
type CreateReducerFn func() Reducer

// Create implements the function of CreateReducer.
func (rf CreateReducerFn) Create() Reducer {
	return rf()
}
```

xxxFn sounds better than xxxFunc

**Review by vigith — APPROVED** [review]

**Review by KeranYang — APPROVED** [review]

File: pkg/reducer/interface.go [hunk 3]
```diff
@@ -23,11 +23,25 @@ type IntervalWindow interface {
 	EndTime() time.Time
 }
 
+// CreateReducer is the interface which is used to create a Reducer.
+type CreateReducer interface {
```
Comment (whynowy) [inline pkg/reducer/interface.go:27]: `CreateReducer` is a verb, is it good for interface name? `ReducerFactory`?

Comment (vigith) [reply pkg/reducer/interface.go:27]: Let's go with `ReducerCreater`?

**Review by KeranYang — APPROVED** [review]

**Comment by KeranYang** [comment]
security check failed, please fix.

File: pkg/reducer/examples/counter/main.go [hunk 4]
```diff
@@ -20,5 +20,7 @@ func reduceCounter(_ context.Context, keys []string, reduceCh <-chan reducer.Dat
 }
 
 func main() {
-	reducer.NewServer(reducer.ReducerFunc(reduceCounter)).Start(context.Background())
+	reducer.NewServer(reducer.ReducerCreatorFn(func() reducer.Reducer {
```
Comment (whynowy) [inline pkg/reducer/examples/counter/main.go:23]: This is too complex to use, can we add a help function like below?
 
```golang
func SimpleReducer(func (_ context.Context, keys []string, reduceCh <-chan reducer.Datum, md reducer.Metadata) reducer.Messages) ReduceCreator {
 ///
}
```

Comment (whynowy) [reply pkg/reducer/examples/counter/main.go:23]:
```golang
type dummyReducerCreator struct {
	f func(ctx context.Context, keys []string, reduceCh <-chan Datum, md Metadata) Messages
}

func (d dummyReducerCreator) Create() Reducer {
	return ReducerFunc(d.f)
}

// NewDummyReducerCreator creates a dummy ReducerCreator.
func NewDummyReducerCreator(f func(ctx context.Context, keys []string, reduceCh <-chan Datum, md Metadata) Messages) ReducerCreator {
	return dummyReducerCreator{f: f}
}
```

Comment (whynowy) [reply pkg/reducer/examples/counter/main.go:23]: `ReducerFunc` can be private in this case.

Comment (whynowy) [reply pkg/reducer/examples/counter/main.go:23]: Start function will be:

```golang
reducer.NewServer(SimpleReducerCreator(reduceCounter)).Start(context.Background())
```

File: pkg/reducer/examples/sum/main.go [hunk 5]
```diff
@@ -9,8 +9,17 @@ import (
 	"github.com/numaproj/numaflow-go/pkg/reducer"
 )
 
+// SumReducer implements the reducer.ReducerCreator interface which creates a reducer
+type SumReducer struct {
```
Comment (whynowy) [inline pkg/reducer/examples/sum/main.go:13]: The struct name is confusing - should not use `xxxReducer`, that sounds like it's an instance of `Reducer` interface.

File: pkg/reducer/examples/counter/main.go (hunk 4 above)
Comment (yhl25) [reply pkg/reducer/examples/counter/main.go:23]: I like this change :)

File: pkg/reducer/examples/sum/main.go (hunk 5 above)
Comment (yhl25) [reply pkg/reducer/examples/sum/main.go:13]: will update it to `SumReducerCreator`

File: pkg/reducer/interface.go [hunk 6]
```diff
@@ -23,15 +23,36 @@ type IntervalWindow interface {
...
+	f func(context.Context, []string, <-chan Datum, Metadata) Messages
+}
+
+// Create creates a Reducer for the given function.
+func (s *simpleReducerCreator) Create() Reducer {
+	return reducerFn(s.f)
+}
+
+// SimpleReducerCreator creates a simple ReducerCreator for the given reduce function.
+func SimpleReducerCreator(f func(context.Context, []string, <-chan Datum, Metadata) Messages) ReducerCreator {
```
Comment (whynowy) [inline pkg/reducer/interface.go:43]: reducer.SimpleCreatorWithReducer(intl)
reducer.SimpleCreatorWithReduceFn(f)

**Review by whynowy — APPROVED** [review]

**Pull request merged**